### PR TITLE
[FLINK-21269] Introduce runtime interfaces for specifying SlotSharingGroup-based resource requirements

### PR DESCRIPTION
**Pull request by KarmaGYZ**
## What is the purpose of the change

Introduce runtime interfaces for specifying SlotSharingGroup-based resource requirements and the specified resource requirements will be passed on all the way to the corresponding SlotSharingGroup in ExecutionGraph.

### CODE REVIEW

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java [hunk 1]
```diff
@@ -1195,6 +1197,55 @@ public void testSlotSharingOnAllVerticesInSameSlotSharingGroupByDefaultDisabled(
...
+        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.fromElements(1, 2, 3)
+                .name(slotSharingGroup1)
+                .slotSharingGroup(slotSharingGroup1)
+                .map(x -> x + 1)
+                .name(slotSharingGroup2)
+                .slotSharingGroup(slotSharingGroup2)
+                .map(x -> x * x)
+                .name(StreamGraphGenerator.DEFAULT_SLOT_SHARING_GROUP)
+                .slotSharingGroup(StreamGraphGenerator.DEFAULT_SLOT_SHARING_GROUP);
```
Comment (xintongsong) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java:1222]: I think we need another test case for an operator that does not specify SSG (thus the default group) while the resource profile of the default group is specified.

**Review by xintongsong — COMMENTED** [review]
Thanks @KarmaGYZ.
The PR LGTM in general. I have only a few minor comment.

**Comment by KarmaGYZ** [comment]
Thanks for the review, @xintongsong . PR updated.

**Review by xintongsong — APPROVED** [review]
LGTM

**Pull request closed without merging**